### DISCUSSION TITLE
Multiply change into Divide

### DISCUSSION THREAD

**Discussion by AlbertoMonteiro** [stryker-mutator/stryker-net#1856] [Q&A] [answered]
I have the following code
```csharp
condition ? request.Amount.Value : request.Amount.Value * -1
```
Styker mutates it to
```csharp
condition ? request.Amount.Value : request.Amount.Value / -1
```

But mathematically, correct if I am wrong, they are the same

![image](https://user-images.githubusercontent.com/836496/148835643-6e928f63-80bb-4e94-860a-ab2330cb5ee7.png)
```csharp
var r = new Random();
Console.WriteLine(string.Join("\n", Enumerable
    .Range(1, 20)
    .Select(_ => r.Next(minValue: -20, maxValue: 20))
    .Select(x => $"Original: {x}, changed: {x * -1} @ {x / -1}")))
```

So, does make sense that mutation be executed in that specific scenario?

**Comment by dupdob** [comment, marked as answer]
Good question.
## Answer
Your question is more theoretical than practical: it does not make sense to test this mutant: if there is no change in the program behavior, this is not an actual mutant, just a source code change.

But, from a practical standpoint, it is normal and expected: Stryker.Net tests every source code mutation it generates. There is a filtering logic that removes mutations that are duplicates **at the source code level**, but there is no attempt to detect semantic duplicates.

## How to avoid ?
When you have mutations that do not change the program behavior, this is a hint the code may be simpler. Here you could use:
```csharp
condition ? request.Amount.Value : -request.Amount.Value
```
## Why Stryker will not detect this?
Adding the necessary logic to detect this specific case (* or / -1) to Stryker will result in a slightly slower mutation generation and a supplemental maintenance burden. Also, this code would still exhibit the same problem:
```
const int MagicFator = - 1;

condition ? request.Amount.Value : request.Amount.Value * MagicFactor
```
## Last resort
Of course, there are situations when the source code cannot be changed in order to prevent semantically equivalent mutant. For those cases, you can prevent the mutation from being tested at the source code level, via Stryker specific comments.

See [Stryker comments](https://github.com/stryker-mutator/stryker-net/blob/master/docs/ignore-mutations.md#stryker-comments).

**Reply by AlbertoMonteiro** [reply]
Thanks for the rich response.
Btw, awesome article: https://many-cores.com/2019/05/17/mutation-testing/

**Reply by dupdob** [reply]
Thank you
